habitat_sim/simulator.py

```diff
@@ -472,24 +448,115 @@ def __del__(self) -> None:
+            render_flags |= habitat_sim.gfx.Camera.Flags.OBJECTS_ONLY
+            self.renderer.draw(
+                sensor_reader._sensor_object,
+                self._sim.get_active_scene_graph(),
+                render_flags,
+            )
+
+    def get_observation(self, sensor_reader: SensorReader) -> Union[ndarray, "Tensor"]:
+
+        tgt = sensor_reader._sensor_object.render_target
+
+        if sensor_reader._spec.gpu2gpu_transfer:
+            with torch.cuda.device(sensor_reader._buffer.device):  # type: ignore[attr-defined]
+                if sensor_reader._spec.sensor_type == SensorType.SEMANTIC:
+                    tgt.read_frame_object_id_gpu(sensor_reader._buffer.data_ptr())  # type: ignore[attr-defined]
+                elif sensor_reader._spec.sensor_type == SensorType.DEPTH:
+                    tgt.read_frame_depth_gpu(sensor_reader._buffer.data_ptr())  # type: ignore[attr-defined]
+                else:
+                    tgt.read_frame_rgba_gpu(sensor_reader._buffer.data_ptr())  # type: ignore[attr-defined]
+
+                obs = sensor_reader._buffer.flip(0)
+        else:
+            size = sensor_reader._sensor_object.framebuffer_size
+
+            if sensor_reader._spec.sensor_type == SensorType.SEMANTIC:
+                tgt.read_frame_object_id(
+                    mn.MutableImageView2D(
+                        mn.PixelFormat.R32UI, size, sensor_reader._buffer
+                    )
+                )
+            elif sensor_reader._spec.sensor_type == SensorType.DEPTH:
+                tgt.read_frame_depth(
+                    mn.MutableImageView2D(
+                        mn.PixelFormat.R32F, size, sensor_reader._buffer
+                    )
+                )
+            else:
+                tgt.read_frame_rgba(
+                    mn.MutableImageView2D(
+                        mn.PixelFormat.RGBA8_UNORM,
+                        size,
+                        sensor_reader._buffer.reshape(
+                            sensor_reader._spec.resolution[0], -1
+                        ),
+                    )
+                )
+
+            obs = np.flip(sensor_reader._buffer, axis=0)
+
+        return sensor_reader._noise_model(obs)
 
-    def __init__(self, sim: Simulator, agent: Agent, sensor_id: str) -> None:
-        self._sim = sim
-        self._agent = agent
 
+class SensorReader:
+    r"""Class to help initialize torch/numpy buffers and noise models for sensors in python"""
+
+    def __init__(self, sim: Simulator, sensor: Sensor) -> None:
+        # Initialize sensor buffers as np arrays and noise models
         # sensor is an attached object to the scene node
         # store such "attached object" in _sensor_object
-        self._sensor_object = self._agent._sensors[sensor_id]
-
-        self._spec = self._sensor_object.specification()
 
-        self._sim.renderer.bind_render_target(self._sensor_object)
+        self._sim = sim
+        self._sensor_object: Sensor
```

Comment:
Ownership is gonna get messy here since the scene graph owns the sensor.  If the sensor reader doesn't keep a reference to the sensor, that'll make it much cleaner

Reply:
Ok, that can be done!